gossip/bloom.go

```diff
@@ -0,0 +1,96 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"encoding/binary"
+	"hash"
+	"time"
+
+	safemath "github.com/ava-labs/avalanchego/utils/math"
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+	"golang.org/x/exp/rand"
+)
+
+var _ hash.Hash64 = (*hasher)(nil)
+
+func NewBloomFilter(m uint64, p float64) (*BloomFilter, error) {
+	bloom, err := bloomfilter.NewOptimal(m, p)
+	if err != nil {
+		return nil, err
+	}
+
+	bloomFilter := &BloomFilter{
+		Bloom: bloom,
+		Salt:  randomSalt(),
+	}
+	return bloomFilter, nil
+}
+
+type BloomFilter struct {
+	Bloom *bloomfilter.Filter
+	// Salt is provided to eventually unblock collisions in Bloom. It's possible
+	// that conflicting Gossipable items collide in the bloom filter, so a salt
+	// is generated to eventually resolve collisions.
+	Salt []byte
+}
+
+func (b *BloomFilter) Add(gossipable Gossipable) {
+	salted := hasher{
+		hash: gossipable.GetHash(),
+		salt: b.Salt,
+	}
+	b.Bloom.Add(salted)
+}
+
+func (b *BloomFilter) Has(gossipable Gossipable) bool {
+	salted := hasher{
+		hash: gossipable.GetHash(),
+		salt: b.Salt,
+	}
+	return b.Bloom.Contains(salted)
+}
+
+// ResetBloomFilterIfNeeded resets a bloom filter if it breaches a ratio of
+// filled elements. Returns true if the bloom filter was reset.
+func ResetBloomFilterIfNeeded(
+	bloomFilter *BloomFilter,
+	maxFilledRatio float64,
+) bool {
+	if bloomFilter.Bloom.PreciseFilledRatio() < maxFilledRatio {
+		return false
+	}
+
+	// it's not possible for this to error assuming that the original
+	// bloom filter's parameters were valid
+	fresh, _ := bloomfilter.New(bloomFilter.Bloom.M(), bloomFilter.Bloom.K())
+	bloomFilter.Bloom = fresh
+	bloomFilter.Salt = randomSalt()
+	return true
+}
+
+func randomSalt() []byte {
+	salt := make([]byte, HashLength)
```

Comment:
I could use a 32 byte hash here. I only made it 8 bytes because `bloomfilter.Filter` uses `Sum64` which is a `uint64`, so only 8 bytes are actually used when using binary encoding.

We could use 32 byte hash for consistency with the rest of the repo.